### ISSUE TITLE
compute_flow.py is reacting to something on a pure black background

### ISSUE THREAD

**Issue by CaptnSeraph** [volotat/SD-CN-Animation#15]
I use runwayml to remove the background from a video, compute_flow renders an odd pattern in the black where it should be seeing nothing, i assume its compression artifacts, but i also ran it through after effects rotobrush to strip anything out and it looks the same, if i change the resolution to 2x the video size then nearly the whole frame is white.

ok ive just rendered it out from after effects into uncompressed avi

![image](https://user-images.githubusercontent.com/29952274/232253740-4fd1c022-cb59-451b-bd69-42e33ae35def.png)

this is a solid black background, am i doing something wrong here?

**Comment by volotat** [comment]
There is nothing wrong on your part. RAFT model is far from perfect, so it tries to predict optical flow where there is none, and the diff map are calculated based on that flow. There are several things you can do to eliminate it. You can go to flow_utils.py and remove occlusion mask from the diff calculation of the 79 line like so:
`alpha_mask = np.maximum(diff_mask_org * 4, diff_mask_stl * 2)`
You may also want to try it like this:
`alpha_mask = diff_mask_org * 4`

Otherwise you could also try a different model that the RAFT method provides. To do so, you would need to change line 19 of flow_utils.py to one of these:
```
'model': 'RAFT/models/raft-sintel.pth'
or
'model': 'RAFT/models/raft-kitti.pth'
or
'model': 'RAFT/models/raft-chairs.pth'
```

If you find out that one of these models works better, please report it here. I haven't tried other models yet.

**Comment by CaptnSeraph** [comment]
ill try the models

**Comment by CaptnSeraph** [comment]
Raft-sintel is arguably worse by far

![image](https://user-images.githubusercontent.com/29952274/232254438-a343ce81-748d-4125-a851-958e14361f7e.png)

raft-kitti is even worse still 

![image](https://user-images.githubusercontent.com/29952274/232254464-7ee8de84-d94b-46cc-946c-7d81a6126335.png)

raft-chairs looked like it was going to be better but then fell apart

![image](https://user-images.githubusercontent.com/29952274/232254497-abad3a89-4f39-4eed-aa6b-9af61a336cea.png)

going back to raft-things but removing the occlusion mask

![image](https://user-images.githubusercontent.com/29952274/232254708-db9fe3d2-5531-4d8a-b719-35bf9e989d03.png)

changing the alpha mask to only use the diff_mask_org * 4

![image](https://user-images.githubusercontent.com/29952274/232254758-f5b92a99-3708-49c4-8731-e05c50214141.png)

i dont see much difference in any method.

is there a better option than using raft? i know it was implemented to reduce VRAM, but is there a more vram expensive method that would yield a better result?

**Comment by volotat** [comment]
This is very strange that you still has white areas with  diff_mask_org * 4. That tells that there are actual differences between frames that become exaggerated because of multipliers. 

Regarding the better flow estimation model, this should work much better: https://github.com/autonomousvision/unimatch but I never worked with this method, so I don't know yet how hard/easy it is to incorporate it to the project.

**Comment by CaptnSeraph** [comment]
I can only assume then that the black frames are somehow filled with noise from the compression method, though i assumed uncompressed AVI was, well... uncompressed.
Ill have to look into the frames themselves and see if i can find some difference in the background, ill also look at implementing unimatch, its about time i got some real coding done

**Comment by CaptnSeraph** [comment]
ive used ffmpeg to pull the frames out as bmp and the black areas are 100% pure black, frame to frame there is zero data in the black areas. 

the only thing i can think is some form of compression

ok, used ffmpeg to put together a file using lib264 and lib264rgb at crf of 0, same thing, there is or should be no variation in the black parts of the image, and yet the optical flow thing is detecting them

**Comment by alexfredo** [comment]
Im not python programmer but maybe background subtraction can help you ? https://www.geeksforgeeks.org/python-background-subtraction-using-opencv/

**Comment by CaptnSeraph** [comment]
i implemented a background removal during the processing (its not a great implementation due to MOG2 being what is, and i dont recommend its use) however it shows the same effect. 

if someone wants to recommend a better method to remove the background im open to listening, it needs testing, but its a start.

https://github.com/volotat/SD-CN-Animation/pull/16

**Comment by alexfredo** [comment]
There's this background remover, I don't know if this better : https://github.com/danielgatis/rembg

**Comment by alexfredo** [comment]
I'm not python programmer but I asked chatgpt for fix the problem and add in flow_utils.py these lines :
 
# remove white noise
    next_flow[np.abs(next_flow) < 3] = 0
    prev_flow[np.abs(prev_flow) < 3] = 0

it seems to have fixed the artifacts problem :
![white](https://user-images.githubusercontent.com/24534698/232312208-f88c2ca2-66e8-44e6-8755-d739104e46be.PNG)

**Comment by alexfredo** [comment]
Here the complete code : 

import numpy as np
import cv2

#RAFT dependencies
import sys
sys.path.append('RAFT/core')

from collections import namedtuple
import torch
import argparse
from raft import RAFT
from utils.utils import InputPadder

RAFT_model = None
def RAFT_estimate_flow(frame1, frame2, device = 'cuda'):
  global RAFT_model
  if RAFT_model is None:
    args = argparse.Namespace(**{
      'model': 'RAFT/models/raft-things.pth',
      'mixed_precision': True,
      'small': False,
      'alternate_corr': False,
      'path': ""
    })

    RAFT_model = torch.nn.DataParallel(RAFT(args))
    RAFT_model.load_state_dict(torch.load(args.model))

    RAFT_model = RAFT_model.module
    RAFT_model.to(device)
    RAFT_model.eval()

  with torch.no_grad():
    frame1_torch = torch.from_numpy(frame1).permute(2, 0, 1).float()[None].to(device)
    frame2_torch = torch.from_numpy(frame2).permute(2, 0, 1).float()[None].to(device)

    padder = InputPadder(frame1_torch.shape)
    image1, image2 = padder.pad(frame1_torch, frame2_torch)

    # estimate optical flow
    _, next_flow = RAFT_model(image1, image2, iters=20, test_mode=True)
    _, prev_flow = RAFT_model(image2, image1, iters=20, test_mode=True)

    next_flow = next_flow[0].permute(1,2,0).cpu().numpy()
    prev_flow = prev_flow[0].permute(1,2,0).cpu().numpy()

    # remove white noise
    next_flow[np.abs(next_flow) < 3] = 0
    prev_flow[np.abs(prev_flow) < 3] = 0

    fb_flow = next_flow + prev_flow
    fb_norm = np.linalg.norm(fb_flow, axis=2)

    occlusion_mask = fb_norm[..., None].repeat(3, axis = -1)

  return next_flow, prev_flow, occlusion_mask

def compute_diff_map(next_flow, prev_flow, prev_frame, cur_frame, prev_frame_styled):
  h, w = cur_frame.shape[:2]

  next_flow = cv2.resize(next_flow, (w, h))
  prev_flow = cv2.resize(prev_flow, (w, h))

  flow_map = -next_flow.copy()
  flow_map[:,:,0] += np.arange(w)
  flow_map[:,:,1] += np.arange(h)[:,np.newaxis]

  warped_frame = cv2.remap(prev_frame, flow_map, None, cv2.INTER_NEAREST)
  warped_frame_styled = cv2.remap(prev_frame_styled, flow_map, None, cv2.INTER_NEAREST)

  # compute occlusion mask
  fb_flow = next_flow + prev_flow
  fb_norm = np.linalg.norm(fb_flow, axis=2)

  occlusion_mask = fb_norm[..., None] 

  diff_mask_org = np.abs(warped_frame.astype(np.float32) - cur_frame.astype(np.float32)) / 255
  diff_mask_org = diff_mask_org.max(axis = -1, keepdims=True)

  diff_mask_stl = np.abs(warped_frame_styled.astype(np.float32) - cur_frame.astype(np.float32)) / 255
  diff_mask_stl = diff_mask_stl.max(axis = -1, keepdims=True)

  alpha_mask = np.maximum(occlusion_mask * 0.3, diff_mask_org * 4, diff_mask_stl * 2)
  alpha_mask = alpha_mask.repeat(3, axis = -1)

  #alpha_mask_blured = cv2.dilate(alpha_mask, np.ones((5, 5), np.float32))
  alpha_mask = cv2.GaussianBlur(alpha_mask, (51,51), 5, cv2.BORDER_DEFAULT)

  alpha_mask = np.clip(alpha_mask, 0, 1)

  return alpha_mask, warped_frame_styled

**Comment by alexfredo** [comment]
The complete file https://www.file.io/q5f2/download/m3VrmDDImSZM

**Comment by alexfredo** [comment]
I need to test on multiples video before for be sure that works well everywhere, also I never made a pull-request but it doesn't seems complicated I can try ^^;

**Comment by alexfredo** [comment]
I noticed something weird with my modification when I generate the flow.h5 file the flow preview show a bad result, but then when I use the flow file with "py vid2vid.py" the preview show a good result without artifacts

**Comment by CaptnSeraph** [comment]
Is that in addition to the background remover code I suggested or for the original method?

I hadn't considered just removing the noise lol.

**Comment by alexfredo** [comment]
The original method

**Comment by alexfredo** [comment]
The problem is when there's a sudden change of scene the accuracy of the motions become bad, chatgpt said "To improve the accuracy of flow estimation when sudden scene changes occur, one approach is to use post-processing techniques such as filtering and smoothing. " is it true ? the other solution and to separate all the scenes, it require more work...

**Comment by CaptnSeraph** [comment]
I know when I'm doing AI frame interpolation you get bad data around scene changes, it passes the sniff test.

But if a little white noise removal is all that's needed then sure why not, the noise is unlikely to be consistent frame to frame but motion usually lasts more than one frame so noise removal makes sense.

**Comment by CaptnSeraph** [comment]
hey, so, funny story... ive been doing some work on this over the last 2 days... made some discoveries...
RAFT shits itself with large areas of solid colour.
If i take a background removed image, and add a checkerboard pattern, most of the noise goes away.
If i add a noise pattern, all of the noise goes away 
so it seems that background removal is not actually doing us any favours on its own.

ive played around with noise removal (btw the noise removal needs to be in the estimate flow part of the script rather than the compute diff map, it seems to help with the noise during both the vizualising and the compute diff part so you can see straight away if its going to give you issues.

as for background removal? i would maybe consider stripping it out tbh, the method i put in is so dirty, i did not expect it to be merged tbh, however if you want to keep it, just move the low value noise suggestion into the estimate flow area

`    padder = InputPadder(frame1_torch.shape)
    image1, image2 = padder.pad(frame1_torch, frame2_torch)

    # estimate optical flow
    _, next_flow = RAFT_model(image1, image2, iters=20, test_mode=True)
    _, prev_flow = RAFT_model(image2, image1, iters=20, test_mode=True)

    next_flow = next_flow[0].permute(1, 2, 0).cpu().numpy()
    prev_flow = prev_flow[0].permute(1, 2, 0).cpu().numpy()
    
    #remove low value noise (@alexfredo suggestion)  <----here
    next_flow[np.abs(next_flow) < 0.1] = 0
    prev_flow[np.abs(prev_flow) < 0.1] = 0

    fb_flow = next_flow + prev_flow`

**Comment by Devicetron** [comment]
if you want true black areas you need to use WebM, HEVC With Alpha or ProRes4444 with transparency enabled. Every other codec will show compression artifacts in black backgrounds.
